jung-api/src/test/java/edu/uci/ics/jung/graph/AbstractCTreeTest.java

```diff
@@ -0,0 +1,588 @@
+ *
+ * All rights reserved.
+ *
+ * This software is open-source under the BSD license; see either
+ * "license.txt" or
+ * https://github.com/jrtom/jung/blob/master/LICENSE for a description.
+ */
+
+package edu.uci.ics.jung.graph;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.common.truth.Truth8.assertThat;
+import static edu.uci.ics.jung.graph.TestUtil.FAIL_ERROR_ELEMENT_NOT_IN_TREE;
+import static edu.uci.ics.jung.graph.TestUtil.FAIL_ERROR_NODEU_NOT_IN_TREE;
+import static edu.uci.ics.jung.graph.TestUtil.FAIL_ERROR_NODEV_IN_TREE;
+import static edu.uci.ics.jung.graph.TestUtil.FAIL_ERROR_TREE_HAS_ROOT;
+import static edu.uci.ics.jung.graph.TestUtil.assertNodeNotInTreeErrorMessage;
+import static edu.uci.ics.jung.graph.TestUtil.assertNodeUNotInTreeErrorMessage;
+import static edu.uci.ics.jung.graph.TestUtil.assertNodeVAlreadyElementOfTreeErrorMessage;
+import static edu.uci.ics.jung.graph.TestUtil.assertStronglyEquivalent;
+import static edu.uci.ics.jung.graph.TestUtil.assertTreeAlreadyHasRootErrorMessage;
+import static edu.uci.ics.jung.graph.TestUtil.sanityCheckSet;
+import static org.junit.Assert.fail;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.graph.EndpointPair;
+import com.google.common.graph.Graphs;
+import com.google.common.graph.ImmutableGraph;
+import java.util.HashSet;
+import java.util.Set;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public abstract class AbstractCTreeTest {
+  MutableCTree<Integer> tree;
+  static final Integer N1 = 1;
+  static final Integer N2 = 2;
+  static final Integer N3 = 3;
+  static final Integer N4 = 4;
+  static final Integer N5 = 5;
+  static final Integer N6 = 6;
+  private static final Integer NODE_NOT_IN_TREE = 1000;
+
+  static final String ERROR_MODIFIABLE_SET = "Set returned is unexpectedly modifiable";
+  static final String ERROR_SELF_LOOP = "self-loops are not allowed";
+  static final String ERROR_ADDED_SELF_LOOP = "Should not be allowed to add a self-loop edge.";
+
+  /** Creates and returns an instance of the tree to be tested. */
+  public abstract MutableCTree<Integer> createTree();
+
+  /**
+   * A proxy method that adds the node {@code n} to the tree being tested.
+   *
+   * @return {@code true} iff the tree was modified as a result of this call
+   */
+  protected boolean addNode(Integer n) {
+    return tree.addNode(n);
+  }
+
+  /**
+   * A proxy method that adds the edge {@code e} to the tree being tested.
+   *
+   * @return {@code true} iff the tree was modified as a result of this call
+   */
+  protected boolean putEdge(Integer n1, Integer n2) {
+    return tree.putEdge(n1, n2);
+  }
+
+  @Before
+  public void init() {
+    tree = createTree();
+  }
+
+  @After
+  public void validateTreeState() {
+    validateTree(tree);
+  }
+
+  static <N> void validateTree(CTree<N> tree) {
+    // TODO: Replace Graphs#copyOf with a tree-specific copyOf method if it ever turns up
```

Comment:
I'm not sure what might cause us to write a tree-specific version of this method.  Did you have anything specific in mind, i.e., how might that be useful?